src/gui/qgsdigitizingguidemaptool.cpp

```diff
@@ -0,0 +1,561 @@
+  connect( mUserInputWidget, &QgsDigitizingGuideToolUserInputWidget::editingFinished, this, &QgsDigitizingGuideMapToolLineAbstract::validateFromUserWidget );
+  connect( mUserInputWidget, &QgsDigitizingGuideToolUserInputWidget::editingCanceled, this, &QgsDigitizingGuideMapToolLineAbstract::restorePreviousMapTool );
+}
+
+void QgsDigitizingGuideMapToolLineAbstract::deleteUserInputWidget()
+{
+  if ( mUserInputWidget )
+  {
+    disconnect( mUserInputWidget, &QgsDigitizingGuideToolUserInputWidget::offsetChanged, this, &QgsDigitizingGuideMapToolLineAbstract::offsetChanged );
+    disconnect( mUserInputWidget, &QgsDigitizingGuideToolUserInputWidget::editingFinished, this, &QgsDigitizingGuideMapToolLineAbstract::validateFromUserWidget );
+    disconnect( mUserInputWidget, &QgsDigitizingGuideToolUserInputWidget::editingCanceled, this, &QgsDigitizingGuideMapToolLineAbstract::restorePreviousMapTool );
+    mUserInputWidget->releaseKeyboard();
+  }
+  mUserInputWidget = nullptr;
+  if ( mFloatingWidget )
+  {
+    mFloatingWidget->deleteLater();
+    mFloatingWidget = nullptr;
+  }
+}
+
+void QgsDigitizingGuideMapToolLineAbstract::offsetChanged( double offset )
+{
+  QgsLineString *line = createLine( QgsPoint(), offset );
+  updateRubberBand( line );
+}
+
+void QgsDigitizingGuideMapToolLineAbstract::validateFromUserWidget( double offset )
+{
+  QgsLineString *line = createLine( QgsPoint(), offset );
+  if ( line )
+  {
+    QgsDigitizingGuideLayer *dl = QgsProject::instance()->digitizingGuideLayer();
+    dl->addLineGuide( line, mUserInputWidget->title() );
+  }
+  restorePreviousMapTool();
+}
+
+QgsDigitizingGuideMapToolLineExtension::QgsDigitizingGuideMapToolLineExtension( QgsMapCanvas *canvas )
+  : QgsDigitizingGuideMapToolLineAbstract( tr( "Line Extension" ), canvas )
+{
+}
+
+QgsLineString *QgsDigitizingGuideMapToolLineExtension::createLine( const QgsPointXY &point, double offset )
+{
+  Q_UNUSED( offset )
+
+  if ( !mSegment )
+    return nullptr;
+
+  double d1 = mSegment->first.sqrDist( point );
+  double d2 = mSegment->second.sqrDist( point );
+
+  const QgsPointXY &startPoint = d1 < d2 ? mSegment->first : mSegment->second;
+
+  const QgsPointXY endPoint = QgsGeometryUtils::perpendicularSegment( QgsPoint( point ), QgsPoint( mSegment->first ), QgsPoint( mSegment->second ) ).endPoint();
+
+  return new QgsLineString( {startPoint, endPoint} );
+}
+
+
+QgsDigitizingGuideMapToolLineParallel::QgsDigitizingGuideMapToolLineParallel( QgsMapCanvas *canvas )
+  : QgsDigitizingGuideMapToolLineAbstract( tr( "Parallel" ), canvas )
+{
+  mHasOffset = true;
+}
+
+QgsLineString *QgsDigitizingGuideMapToolLineParallel::createLine( const QgsPointXY &point, double offset )
+{
+  if ( !mSegment )
+    return nullptr;
+
+  const QgsPointXY &s1 = mSegment->first;
+  const QgsPointXY &s2 = mSegment->second;
+
+  const double nx = s2.y() - s1.y();
```

Comment:
Some high level comment about what all the below maths is trying to accomplish would be welcome :-)
I'd note that there's a division by zero if s1 == s2

Wondering if there wouldn't be value to move the computation part to some non-GUI method in case it might be helpful (but no strong opinion)